### PR TITLE
Update substrate, add block construction metrics

### PR DESCRIPTION
**Pull request by NikVolf**

### CODE REVIEW

File: Cargo.lock [hunk 1]
```diff
@@ -4163,7 +4145,7 @@ dependencies = [
  "sp-core",
  "sp-keyring",
  "sp-runtime",
- "tokio 0.2.21",
+ "tokio 0.2.19",
```
Comment (bkchr) [inline Cargo.lock:4148]: Why does this downgrade that many packages?

File: validation/src/block_production.rs [hunk 2]
```diff
@@ -175,6 +185,8 @@ impl<Client, TxPool, Backend> consensus::Proposer<Block> for Proposer<Client, Tx
 		let transaction_pool = self.transaction_pool.clone();
 		let table = self.tracker.table().clone();
 		let backend = self.backend.clone();
+		let metrics = self.metrics.clone();
+		let block_timer = metrics.report(|metrics| metrics.block_constructed.start_timer());
```
Comment (bkchr) [inline validation/src/block_production.rs:189]: This should be moved into the `async move`. AFAIK the async code gets called on the first poll?

File: validation/src/block_production.rs [hunk 3]
```diff
@@ -214,11 +226,18 @@ impl<Client, TxPool, Backend> consensus::Proposer<Block> for Proposer<Client, Tx
...
+			let result = tokio::task::spawn_blocking(
+				move || {
+					let proposed_candidates = table.proposed_set();
+					data.propose_with(proposed_candidates)
+				}
+			).await?;
+
+			drop(block_timer);
+			if let Ok(proposal) = result.as_ref() {
+				metrics.report(|metrics| metrics.number_of_transactions.set(proposal.block.extrinsics.len() as u64));
```
Comment (bkchr) [inline validation/src/block_production.rs:238]: Hmmm. We should add a new metric for "failed" proposals. Or what do you think?

Comment (bkchr) [reply validation/src/block_production.rs:238]: Or just use `0`

File: Cargo.lock (hunk 1 above)
Comment (NikVolf) [reply Cargo.lock:4148]: Yeah, no idea

I'll reset Cargo.lock

File: validation/src/block_production.rs (hunk 2 above)
Comment (NikVolf) [reply validation/src/block_production.rs:189]: Yeah, but I doubt this will make much difference

**Pull request merged**